### DISCUSSION TITLE
Lesson 7: Cheaper withdraw continued - FAIL Reason: Assertion Violated

### DISCUSSION THREAD

**Discussion by 0xTagger** [Cyfrin/foundry-full-course-cu#603] [Q&A] [answered]
Hi all - I am having trouble completing video 99. in Lesson 7.

I an receiving an error when using "forge snapshot" when I try to implement and test the testWithdrawFromMultipleFundersCheaper() function. I'm really not sure what I am missing and I can not find much info elsewhere related to the issue.

Any feedback is much appreciated - thank you! 

<img width="815" alt="Screenshot 2023-08-11 at 10 00 14 AM" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/114359465/4c08671b-59b3-40c3-a85e-3fb73ea958bc">
<img width="1083" alt="Screenshot 2023-08-11 at 9 53 43 AM" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/114359465/7ee0a27d-2222-43d3-82df-193fa6de0e2a">

**Comment by shikhar229169** [comment]
Can you run the below command and provide the logs:
```
forge test --mt <enter_your_test_function_name_here> -vvvv
```

**Reply by 0xTagger** [reply]
<img width="840" alt="Screenshot 2023-08-11 at 11 45 43 AM" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/114359465/da8e8339-a04c-4716-9a34-0d3d46d2c11c">
<img width="831" alt="Screenshot 2023-08-11 at 11 46 45 AM" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/114359465/31054de3-da7e-4406-849d-b1b0f273133e">
<img width="641" alt="Screenshot 2023-08-11 at 11 47 05 AM" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/114359465/db16eea2-1d60-4c9d-9d61-1e6ed9e171d1">

**Reply by shikhar229169** [reply, marked as answer]
The problem seems to be with cheaperWithdraw function. I think in the cheaperWithdraw function you are not sending eth to owner, check that.

**Reply by 0xTagger** [reply]
Thank you @shikhar229169 It was in fact an issue with the cheaperWithdraw function! Although I am confused on what the exact issue was if you could provide a further explanation it would be much appreciated.

My Original FundMe.sol (cheaperWithdraw function) code that did NOT work:
<img width="403" alt="Screenshot 2023-08-11 at 12 21 03 PM" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/114359465/96678f2e-ddd7-4edf-b02b-f6c22a19a0bb">

Cyfrin Repo FundMe.sol (cheaperWithdraw function) code that did work:
<img width="645" alt="Screenshot 2023-08-11 at 12 20 24 PM" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/114359465/28bde424-4c46-4ecb-a8e5-d7ccf26f8912">

What was missing in my original code (the memory variable and the bool at the bottom) was not in the video lesson up to this point so I am confused on how I was missing it while following along, but is there an explanation as to why when adding these to the function that it was able to pass the test properly compared to what I had input originally? What is the logic behind it?

**Reply by shikhar229169** [reply]
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/111875681/eea577f2-f482-4e29-ac06-248a57d87a07)

You have not added this line, which sends the eth to the owner.
And as eth was not sent to the owner, so the contract balance was not 0.

**Reply by cell0901** [reply]
i am confused, then why did his code work when he removed the line you mentioned from withdraw() func?
